 We are going to work on 'Customer purchases history' dataset provided by Kaggle

Dataset description: This dataset contains historical records for customer purchases from a supermarket X. The objective is to find the association rules to help supermarket owners find new marketing plans to improve their sales.


Instructions

toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

      ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

Run the apriori algorithm on the provided toy_dataset. Interpret the results.

Try to explore the checkpoint dataset using Pandas and Plotly.

Run the apriori algorithm on checkpoint dataset. Interpret the results and suggest a clear business plan to the supermarket owners based on your findings.



In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori


In [2]:
# Toy dataset
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],
               ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],
               ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],
               ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],
               ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [4]:
te= TransactionEncoder()
te_ary= te.fit(toy_dataset).transform(toy_dataset) #To fit and encode the dataset
encoded_dataset= pd.DataFrame(te_ary, columns=te.columns_) #To return a dataframe of the encoded dataset
encoded_dataset

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [5]:
support = apriori(encoded_dataset)
support #Put in the dataset in apriori
#It returns the support of each of the items or itemset. The items names are the indicies

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support itemsets
0      0.6      (2)
1      0.8      (5)
2      0.8      (7)
3      0.6      (9)
4      0.6   (5, 7)
5      0.6   (9, 5)

In [6]:
#to show the item names and the support
support_new = apriori(encoded_dataset,use_colnames=True)
support_new

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support        itemsets
0      0.6           (Hat)
1      0.8         (Pants)
2      0.8         (Scarf)
3      0.6         (Skirt)
4      0.6  (Pants, Scarf)
5      0.6  (Pants, Skirt)

In [7]:
#for confidence and lift
from mlxtend.frequent_patterns import association_rules
association_rules(support_new, metric="confidence") # confidence - fit it in the dataset that that the support and itemset

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  support  \
0     (Skirt)     (Pants)                 0.6                 0.8      0.6   

   confidence  lift  leverage  conviction  zhangs_metric  
0         1.0  1.25      0.12         inf            0.5

In [8]:
#for lift
association_rules(support_new,metric="lift")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/pandas/core/nanops.py:1010: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


antecedents consequents  antecedent support  consequent support  support  \
0     (Pants)     (Scarf)                 0.8                 0.8      0.6   
1     (Scarf)     (Pants)                 0.8                 0.8      0.6   
2     (Pants)     (Skirt)                 0.8                 0.6      0.6   
3     (Skirt)     (Pants)                 0.6                 0.8      0.6   

   confidence    lift  leverage  conviction  zhangs_metric  
0        0.75  0.9375     -0.04         0.8          -0.25  
1        0.75  0.9375     -0.04         0.8          -0.25  
2        0.75  1.2500      0.12         1.6           1.00  
3        1.00  1.2500      0.12         inf           0.50

Interpretation of result

There is a high support for hats, pants, scarfs and skirts of all the items in the store.

The confidence for Pants and Skirts is high

The lift for pants and skirt is 1.25 which means the two itmes go together. There is a high chance of a customer buy both items at the same time.

In [13]:
Market = pd.read_csv('/content/drive/MyDrive/Market_Basket_Optimisation.csv', header = None)
Market.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [22]:
#To convert the dataframe to a nested list

Market_list = Market.values.tolist()
Market_list

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers',
  'meatballs',
  'eggs',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan],
 ['chutney',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan],
 ['turkey',
  'avocado',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan],
 ['mineral water',
  'milk',
  'energy bar',
  'whole wheat rice',
  'green tea',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan],
 ['low fat yogurt',
  nan,
  n

In [15]:
#To remove the Nan in the data
Market_list2 = [[x for x in y if pd.notna(x)] for y in Market_list]
Market_list2

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spagh

In [17]:
#Encode the new data and convert it to a dataframe
tr_en = TransactionEncoder()
Market_encoded =tr_en.fit(Market_list2).transform(Market_list2)
Market_encoded = pd.DataFrame(Market_encoded, columns=tr_en.columns_)
Market_encoded.head()


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

In [20]:
#To get the support fit the encoded data set into apriori
support_1 = apriori(Market_encoded, min_support= 0.01)
support_1.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support itemsets
0  0.020397      (1)
1  0.033329      (4)
2  0.010799      (7)
3  0.014265      (8)
4  0.011465     (10)

In [23]:
Support_2 = apriori(Market_encoded, min_support=0.01, use_colnames=True)
Support_2.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support          itemsets
0  0.020397         (almonds)
1  0.033329         (avocado)
2  0.010799  (barbecue sauce)
3  0.014265       (black tea)
4  0.011465      (body spray)

In [24]:
#for confidence
association_rules(Support_2, min_threshold= 0.01, metric="confidence")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                 consequents  \
0              (mineral water)                   (avocado)   
1                    (avocado)             (mineral water)   
2                    (burgers)                      (cake)   
3                       (cake)                   (burgers)   
4                  (chocolate)                   (burgers)   
..                         ...                         ...   
427      (spaghetti, pancakes)             (mineral water)   
428  (mineral water, pancakes)                 (spaghetti)   
429                (spaghetti)   (mineral water, pancakes)   
430            (mineral water)       (spaghetti, pancakes)   
431                 (pancakes)  (spaghetti, mineral water)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.238368            0.033329  0.011598    0.048658  1.459926   
1              0.033329            0.238368  0.011598    0.348000  1.459926   
2              0.087188            0.081056  0.011465    0.131498  1.622319   
3              0.081056            0.087188  0.011465    0.141447  1.622319   
4              0.163845            0.087188  0.017064    0.104150  1.194537   
..                  ...                 ...       ...         ...       ...   
427            0.025197            0.238368  0.011465    0.455026  1.908923   
428            0.033729            0.174110  0.011465    0.339921  1.952333   
429            0.174110            0.033729  0.011465    0.065850  1.952333   
430            0.238368            0.025197  0.011465    0.048098  1.908923   
431            0.095054            0.059725  0.011465    0.120617  2.019529   

     leverage  conviction  zhangs_metric  
0    0.003654    1.016113       0.413630  
1    0.003654    1.168147       0.325896  
2    0.004398    1.058080       0.420238  
3    0.004398    1.063198       0.417434  
4    0.002779    1.018933       0.194767  
..        ...         ...            ...  
427  0.005459    1.397557       0.488452  
428  0.005593    1.251198       0.504819  
429  0.005593    1.034385       0.590626  
430  0.005459    1.024059       0.625163  
431  0.005788    1.069244       0.557862  

[432 rows x 10 columns]

In [25]:
#for lift
association_rules(Support_2,min_threshold= 1, metric="lift")


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents                 consequents  \
0              (mineral water)                   (avocado)   
1                    (avocado)             (mineral water)   
2                    (burgers)                      (cake)   
3                       (cake)                   (burgers)   
4                  (chocolate)                   (burgers)   
..                         ...                         ...   
401      (spaghetti, pancakes)             (mineral water)   
402  (mineral water, pancakes)                 (spaghetti)   
403                (spaghetti)   (mineral water, pancakes)   
404            (mineral water)       (spaghetti, pancakes)   
405                 (pancakes)  (spaghetti, mineral water)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.238368            0.033329  0.011598    0.048658  1.459926   
1              0.033329            0.238368  0.011598    0.348000  1.459926   
2              0.087188            0.081056  0.011465    0.131498  1.622319   
3              0.081056            0.087188  0.011465    0.141447  1.622319   
4              0.163845            0.087188  0.017064    0.104150  1.194537   
..                  ...                 ...       ...         ...       ...   
401            0.025197            0.238368  0.011465    0.455026  1.908923   
402            0.033729            0.174110  0.011465    0.339921  1.952333   
403            0.174110            0.033729  0.011465    0.065850  1.952333   
404            0.238368            0.025197  0.011465    0.048098  1.908923   
405            0.095054            0.059725  0.011465    0.120617  2.019529   

     leverage  conviction  zhangs_metric  
0    0.003654    1.016113       0.413630  
1    0.003654    1.168147       0.325896  
2    0.004398    1.058080       0.420238  
3    0.004398    1.063198       0.417434  
4    0.002779    1.018933       0.194767  
..        ...         ...            ...  
401  0.005459    1.397557       0.488452  
402  0.005593    1.251198       0.504819  
403  0.005593    1.034385       0.590626  
404  0.005459    1.024059       0.625163  
405  0.005788    1.069244       0.557862  

[406 rows x 10 columns]

Interpretation of Result

There is a high level of lift among most of the items. The supermarket should stock up on the items.